pkg/validator/runner.go

```diff
@@ -0,0 +1,220 @@
+package validator
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/mt-sre/addon-metadata-operator/pkg/types"
+	"github.com/mt-sre/addon-metadata-operator/pkg/utils"
+)
+
+var initializers []Initializer
+
+// Register queues an "Initializer" which the "Runner"
+// will invoke upon startup.
+func Register(init Initializer) {
+	initializers = append(initializers, init)
+}
+
+// Initializer is a function which will initialize a Validator
+// with dependencies. An error is returned if the Validator
+// cannot be initalized properly.
+type Initializer func(Dependencies) (Validator, error)
```

Comment:
Is this the replacement for middlewares? How can we migrate the retry-logic middleware? Or inject a prometheus reporter middleware for each validator?

Reply:
Not really. Middleware can still be added with a signature like `type Middleware func(Validator) Validator` and any Middleware would be injected as a `RunnerOption`. The `Runner` would refer to a `[]Middleware` during its `Run` function to wrap each Validator. That way you can have pre/post task execution like timers or metric collection. Anything more granular like debug logging would be injected from the `Initializer` func.